File: src/main/java/com/alpha/problemsolving/DailyCodingProblem8.java
package com.alpha.problemsolving;

/*
A unival tree (which stands for "universal value") is a tree where all nodes under it have the same value.

Given the root to a binary tree, count the number of unival subtrees.

For example, the following tree has 5 unival subtrees:

   0
  / \
 1   0
    / \
   1   0
  / \
 1   1

*/

class DailyCodingProblem8 {

    public static void main(String args[]) {

        BinaryTree tree = new BinaryTree();
        tree.root = new Node(0, new Node(1), new Node(0, new Node(1, new Node(1), new Node(1)), new Node(0)));

        int res = tree.countUnivalTrees(tree.root);
        System.out.println(res);

        /*
              5
             / \
            4   5
           / \   \
          4   4   5

        */

        tree = new BinaryTree();
        tree.root = new Node(5);
        tree.root.left = new Node(4);
        tree.root.right = new Node(5);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(4);
        tree.root.right.right = new Node(5);
        res = tree.countUnivalTrees(tree.root);
        System.out.println(res);

    }

}

class Node {
    public int value;
    public Node left, right;

    Node(int value, Node left, Node right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    Node(int value) {
        this.value = value;
        this.left = this.right = null;
    }
}

class BinaryTree {
    Node root;

    int countUnivalTrees(Node root) {
        if (root == null) {
            return 0;
        }
        int count = countUnivalTrees(root.left) + countUnivalTrees(root.right);
        if (root.left != null && root.value != root.left.value) {
            return count;
        }

        if (root.right != null && root.value != root.right.value) {
            return count;
        }

        // if whole tree is unival tree
        return count + 1;
    }

}

File: src/main/typescript/DailyCodingProblem8.ts
/*
This problem was asked by Google.

A unival tree (which stands for "universal value") is a tree where all nodes under it have the same value.

Given the root to a binary tree, count the number of unival subtrees.

For example, the following tree has 5 unival subtrees:

   0
  / \
 1   0
    / \
   1   0
  / \
 1   1

*/

class TreeNode {
    _val: number;
    _left: TreeNode;
    _right: TreeNode;
    constructor(val: number, left: TreeNode = null, right: TreeNode = null) {
        this._val = val;
        this._left = left;
        this._right = right;
    }
}


function countUniversalTreeNodes(root: TreeNode): [boolean, number] {
    if (root === null) return [true, 0];
    //count universal nodes on left
    const [is_left_unival, left_count] = countUniversalTreeNodes(root._left);
    //count universal nodes on right
    const [is_right_unival, right_count] = countUniversalTreeNodes(root._right);
    //If both left and right trees are universal
    if (is_left_unival && is_right_unival) {
        //If left val is not equal to root val then retrun false
        if (root._left != null && root._left._val != root._val) {
            return [false, left_count + right_count];
        }
        if (root._right != null && root._right._val != root._val) {
            return [false, left_count + right_count];
        }
        /*
        This is not needed as one of above will true if root._left._val != root._right._val
        if (root._left._val != root._right._val) {
        return [false, left_count + right_count];}
                */
        return [true, left_count + right_count + 1]
    }
    return [false, left_count + right_count];
}

let root: TreeNode = new TreeNode(0, new TreeNode(1), new TreeNode(0, new TreeNode(1, new TreeNode(1), new TreeNode(1)), new TreeNode(0)))

console.log(countUniversalTreeNodes(root));
/*
   1
  / \
 1   1
    / \
   1   1
      / \
     1   0


*/

root = new TreeNode(1, new TreeNode(1, new TreeNode(1), new TreeNode(1,new TreeNode(1), new TreeNode(0))), new TreeNode(1));

console.log(countUniversalTreeNodes(root));